woof/code/cannon-blast-code.html

```diff
@@ -0,0 +1,323 @@
+        hpBar.x -= 3;
+        
+        // change the color to yellow if 50% or less
+        if (hpBar.width < 112 && hpBar.width > 56) {
+          hpBar.color = "#D2C533";
+        }
+        
+        // change the color to red if 25% or less
+        else if (hpBar.width < 56) {
+          hpBar.color = "#DB5D67";
+        }
+        
+        // end the game if it gets too small
+        if (hpBar.width < 3) {
+          hpBar.hide();
+          freeze();
+        }
+      }
+    }
+  });
+});
+
+
+// player shooting
+// set the bullet's angle to the same as the cannon so it shoots the right way
+// move the bullet so it lines up with the tip of the cannon
+var playerBullets = [];
+onKeyDown(() => {
+  if (keysDown.includes('space')) {
+    var wpn = new Circle({
+      radius: 5,
+      color: "#5D67DB",
+      angle: cannon.angle
+    });
+    wpn.move(29);
+    playerBullets.push(wpn);
+  }
+});
+
+var enemies = [];
+// pass a function to every() to create enemies more quickly as the time goes up
+every(
+  () => {
+    if (time < 60) 
+      return random(1.5, 2.8);
+    else if (time < 90)
+      return random(0.3, 1.5);
+    else if (time < 130)
+      return random(0.1, 0.7);
+  }, 
+  "second", 
+  () => {
+    var enemy = new Rectangle({
+      width: 15,
+      height: 15,
+      color: "#AF4042",
+      x: maxX,
+      y: -100
+    });
+    enemies.push(enemy);
+  }
+);
+
+// create tanks after 30 seconds
+// as the time increases, create them faster
+var tanks = [];
+every(
+  () => {
+    if (time < 75)
+      return random(4.55, 6.01);
+    else if (time < 105)
+      return random(2.55, 4.55);
+    else if (time < 120)
+      return random(1.5, 2.5);
+    else if (time < 150)
+      return random(0.75, 1.5);
+  }, 
+  "second", 
+  () => {
+    if (time >= 30) {
+  		var tank = new Rectangle({
+  			width: 25,
+        height: 25,
+        color: "#40AFAD",
+        x: maxX,
+        y: -100
+  		});
+  		tank.health = 3;
+  		tanks.push(tank);
+  	}
+  }
+);
+
+// another way to have the enemies spawn randomly and increase frequency
+// would be to follow a pattern like this:
+// every(() => random(1.5, 2.8), 'second', () => {
+//   if (time < 60) {
+//     var enemy = new Rectangle({
+//     width: 15,
+//     height: 15,
+//     color: "#AF4042",
+//     x: maxX,
+//     y: -100
+//   })
+//   enemies.basic.push(enemy)
+//   }
+// })
+// every(() => random(0.3, 1.5), 'second', () => {
+//   if (time > 60 && time < 90) {
+//     var enemy = new Rectangle({
+//     width: 15,
+//     height: 15,
+//     color: "#AF4042",
+//     x: maxX,
+//     y: -100
+//   })
+//   enemies.basic.push(enemy)
+//   }
+// })
+//
+// etc. etc. for each increase
+
+
+
+// have the enemies regularly shoot their weapons if they are close enough
+var enemyWeapons = [];
+every(2, 'second', () => {
+  enemies.forEach(enemy => {
+    if (enemy.showing && enemy.shoot) {
```

Comment:
Given that you're going to delete `enemy` from the enemies array on deletion, you probably don't need to check if they're showing here. Also, instead of calculating `.shoot` somewhere, else you could just put `enemy.distanceTo(baseMain) < 150` in this `if`.